cmd/clusterctl/pkg/client/cluster/upgrader.go

```diff
@@ -0,0 +1,275 @@
+		// This is driving all the plan logic, because all the providers in a management group are expected to be consistent
+		// with the Cluster API version supported by the core provider. e.g if the core provider is v1alpha3, all the provider in the same management group should be v1alpha3 as well.
+		coreProviderVersion, err := version.ParseSemantic(managementGroup.CoreProvider.Version)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse current version for the %s/%s provider", managementGroup.CoreProvider.Namespace, managementGroup.CoreProvider.Name)
+		}
+
+		coreProviderMetadata, _, err := u.getUpgradeInfo(managementGroup.CoreProvider)
+		if err != nil {
+			return nil, err
+		}
+
+		// Identifies ClusterAPI version supported by the core provider; this includes the current Cluster API version (e.g. v1alpha3) and the new one available, if any.
+		clusterAPIVersionForUpgrade := getClusterAPIVersionForUpgrade(*coreProviderVersion, coreProviderMetadata)
+
+		if len(clusterAPIVersionForUpgrade) == 0 {
+			return nil, errors.Wrapf(err, "Invalid metadata: unable to find current ClusterAPIVersion for the %s/%s provider", managementGroup.CoreProvider.Namespace, managementGroup.CoreProvider.Name)
+		}
+
+		// Creates UpgradePlans for each cluster API
+		for _, clusterAPIVersion := range clusterAPIVersionForUpgrade {
+			plan := UpgradePlan{
+				ClusterAPIVersion: clusterAPIVersion,
+			}
+
+			// Identifies the upgrade info for each provider
+			// TODO: refactor in order to avoid the nested loop
+			for _, provider := range managementGroup.Providers {
+				// Gets the current version of the core provider, the provider metadata with the link between release series and Cluster API versions, and the list of versions available for upgrades.
+				providerVersion, err := version.ParseSemantic(provider.Version)
+				if err != nil {
+					return nil, errors.Wrapf(err, "failed to parse current version for the %s/%s provider", provider.Namespace, provider.Name)
+				}
+
+				providerMetadata, versions, err := u.getUpgradeInfo(provider)
+				if err != nil {
+					return nil, err
+				}
+
+				// Identifies the next UpgradeVersion within the clusterAPIVersion, if available
+				upgradeVersion := getUpgradeVersion(*providerVersion, clusterAPIVersion, versions, providerMetadata)
+
+				plan.Providers = append(plan.Providers, UpgradeItem{Provider: provider, TargetVersion: versionTag(upgradeVersion)})
+			}
+
+			ret = append(ret, plan)
+		}
+	}
+
+	return ret, nil
+}
+
+// getUpgradeInfo returns all the info required for upgrade plan, including the provider metadata with the link between release series and
+// Cluster API versions, and the list of versions available in the provider's repository.
+func (u *upgradeService) getUpgradeInfo(provider clusterctlv1.Provider) (*clusterctlv1.Metadata, []version.Version, error) {
```

Comment:
```suggestion
func (u *upgradeService) getUpgradeInfo(provider clusterctlv1.Provider) (*upgradeInfo, error) {
```

Reply:
Done. This opened up to a nice refactor of the code, so many thanks for this suggestion!